File: lib/email.ts
import { Resend } from "resend";

const resend = process.env.RESEND_API_KEY
  ? new Resend(process.env.RESEND_API_KEY)
  : null;

const fromEmail =
  process.env.RESEND_FROM_EMAIL || "OpenLingo <onboarding@resend.dev>";

interface SendEmailOptions {
  to: string;
  subject: string;
  html: string;
}

export async function sendEmail({ to, subject, html }: SendEmailOptions) {
  if (!resend) {
    console.warn(
      "[email] RESEND_API_KEY is not set — skipping email send to:",
      to
    );
    console.warn("[email] Subject:", subject);
    return;
  }

  const { data, error } = await resend.emails.send({
    from: fromEmail,
    to: [to],
    subject,
    html,
  });

  if (error) {
    console.error("[email] Failed to send email:", error);
    return;
  }

  return data;
}

File: refactor_rust/backend/src/services/email.rs
use crate::{
    error::{AppError, AppResult},
    state::AppState,
};

pub async fn send_html(state: &AppState, to: &str, subject: &str, html: &str) -> AppResult<()> {
    let Some(api_key) = state.config.resend_api_key.as_deref() else {
        tracing::info!(%to, %subject, "RESEND_API_KEY unset; skipping email");
        return Ok(());
    };

    let payload = serde_json::json!({
        "from": state.config.resend_from_email,
        "to": [to],
        "subject": subject,
        "html": html,
    });
    let res = state
        .http
        .post("https://api.resend.com/emails")
        .bearer_auth(api_key)
        .json(&payload)
        .send()
        .await
        .map_err(|e| AppError::External(format!("email request failed: {e}")))?;
    if !res.status().is_success() {
        return Err(AppError::External(format!(
            "email returned {}",
            res.status()
        )));
    }
    Ok(())
}
